src/crimson/os/seastore/lba_manager/btree/lba_btree_node.h

```diff
@@ -136,72 +136,157 @@ struct lba_map_val_le_t {
   }
 };
 
+template <bool val>
+struct leaf_node_type_t {};
+
+template<>
+struct leaf_node_type_t<true> {
+  static constexpr extent_types_t type = extent_types_t::LADDR_LEAF;
+};
+
+template<>
+struct leaf_node_type_t<false> {
+  static constexpr extent_types_t type = extent_types_t::DINK_LADDR_LEAF;
+};
+
+template <bool has_children>
 struct LBALeafNode
   : FixedKVLeafNode<
       LEAF_NODE_CAPACITY,
       laddr_t, laddr_le_t,
       lba_map_val_t, lba_map_val_le_t,
       LBA_BLOCK_SIZE,
-      LBALeafNode> {
+      LBALeafNode<has_children>,
+      has_children> {
   using Ref = TCachedExtentRef<LBALeafNode>;
-  using internal_iterator_t = const_iterator;
+  using parent_type_t = FixedKVLeafNode<
+			  LEAF_NODE_CAPACITY,
+			  laddr_t, laddr_le_t,
+			  lba_map_val_t, lba_map_val_le_t,
+			  LBA_BLOCK_SIZE,
+			  LBALeafNode<has_children>,
+			  has_children>;
+  using internal_const_iterator_t =
+    typename parent_type_t::node_layout_t::const_iterator;
+  using internal_iterator_t =
+    typename parent_type_t::node_layout_t::iterator;
   template <typename... T>
   LBALeafNode(T&&... t) :
-    FixedKVLeafNode(std::forward<T>(t)...) {}
+    parent_type_t(std::forward<T>(t)...) {}
+
+  static constexpr extent_types_t TYPE = leaf_node_type_t<has_children>::type;
 
-  static constexpr extent_types_t TYPE = extent_types_t::LADDR_LEAF;
+  bool validate_stable_children() final {
+    LOG_PREFIX(LBALeafNode::validate_stable_children);
+    if constexpr (has_children) {
+      if (this->children.empty()) {
+	return false;
+      }
+
+      for (auto i : *this) {
+	auto child = (LogicalCachedExtent*)this->children[i.get_offset()];
+	if (is_valid_child_ptr(child)
+	    && !child->is_clean_pending()
+	    && child->get_laddr() != i.get_key()) {
+	  SUBERROR(seastore_fixedkv_tree,
+	    "stable child not valid: child {}, key {}",
+	    *child,
+	    i.get_key());
+	  ceph_abort();
+	  return false;
+	}
+      }
+    }
+    return true;
+  }
 
   void update(
-    const_iterator iter,
-    lba_map_val_t val) final {
-    val.paddr = maybe_generate_relative(val.paddr);
-    return journal_update(
+    internal_const_iterator_t iter,
+    lba_map_val_t val,
+    LogicalCachedExtent* nextent) final {
+    LOG_PREFIX(LBALeafNode::update);
+    if constexpr (has_children) {
+      if (nextent) {
+	SUBTRACE(seastore_fixedkv_tree, "trans.{}, pos {}, {}",
+	  this->pending_for_transaction,
+	  iter.get_offset(),
+	  *nextent);
+	this->update_child_ptr(iter, nextent);
+      }
```

Comment:
Worth a comment:
// child-ptr may already be correct, see LBAManager::update_mappings()